File: Leetcode/0086-partition-list/solution.go
package main

import "fmt"

// 0086. Partition List
type ListNode struct {
	Val  int
	Next *ListNode
}

func partition(head *ListNode, x int) *ListNode {
	lessDummy := &ListNode{}
	geDummy := &ListNode{}
	lt, gt := lessDummy, geDummy
	for cur := head; cur != nil; cur = cur.Next {
		if cur.Val < x {
			lt.Next = cur
			lt = lt.Next
		} else {
			gt.Next = cur
			gt = gt.Next
		}
	}
	gt.Next = nil
	lt.Next = geDummy.Next
	return lessDummy.Next
}

func toList(arr []int) *ListNode {
	if len(arr) == 0 {
		return nil
	}
	h := &ListNode{Val: arr[0]}
	c := h
	for _, v := range arr[1:] {
		c.Next = &ListNode{Val: v}
		c = c.Next
	}
	return h
}
func toSlice(h *ListNode) []int {
	out := []int{}
	for h != nil {
		out = append(out, h.Val)
		h = h.Next
	}
	return out
}
func eq(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func main() {
	passed, failed := 0, 0
	test := func(name string, got, want []int) {
		if eq(got, want) {
			fmt.Printf("PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("FAIL: %s got=%v\n", name, got)
			failed++
		}
	}
	cases := []struct {
		name string
		in   []int
		x    int
		want []int
	}{
		{"Example 1", []int{1, 4, 3, 2, 5, 2}, 3, []int{1, 2, 2, 4, 3, 5}},
		{"Example 2", []int{2, 1}, 2, []int{1, 2}},
		{"Empty", []int{}, 5, []int{}},
		{"Single", []int{5}, 5, []int{5}},
		{"All less", []int{1, 2, 3}, 5, []int{1, 2, 3}},
		{"All greater", []int{5, 6, 7}, 1, []int{5, 6, 7}},
		{"All equal", []int{3, 3, 3}, 3, []int{3, 3, 3}},
		{"Negatives", []int{-3, 1, -1, 0}, 0, []int{-3, -1, 1, 0}},
		{"Mixed", []int{1, 4, 3, 2, 5, 2}, 3, []int{1, 2, 2, 4, 3, 5}},
	}
	for _, c := range cases {
		test(c.name, toSlice(partition(toList(c.in), c.x)), c.want)
	}
	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}

File: Leetcode/0086-partition-list/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        """Time O(n), Space O(1)."""
        less_dummy = ListNode(0)
        ge_dummy = ListNode(0)
        lt = less_dummy
        gt = ge_dummy
        cur = head
        while cur:
            if cur.val < x:
                lt.next = cur; lt = lt.next
            else:
                gt.next = cur; gt = gt.next
            cur = cur.next
        gt.next = None
        lt.next = ge_dummy.next
        return less_dummy.next


def to_list(arr):
    if not arr: return None
    h = ListNode(arr[0]); c = h
    for v in arr[1:]: c.next = ListNode(v); c = c.next
    return h


def to_arr(h):
    out = []
    while h: out.append(h.val); h = h.next
    return out


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    def test(name, got, exp):
        global passed, failed
        if got == exp: print(f"PASS: {name}"); passed += 1
        else: print(f"FAIL: {name} got={got}"); failed += 1
    cases = [
        ("Example 1", [1, 4, 3, 2, 5, 2], 3, [1, 2, 2, 4, 3, 5]),
        ("Example 2", [2, 1], 2, [1, 2]),
        ("Empty", [], 5, []),
        ("Single", [5], 5, [5]),
        ("All less", [1, 2, 3], 5, [1, 2, 3]),
        ("All greater", [5, 6, 7], 1, [5, 6, 7]),
        ("All equal", [3, 3, 3], 3, [3, 3, 3]),
        ("Negatives", [-3, 1, -1, 0], 0, [-3, -1, 1, 0]),
    ]
    for n, a, x, exp in cases: test(n, to_arr(sol.partition(to_list(a), x)), exp)
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
